server/controllers/caseNotesController.ts

```diff
@@ -20,63 +29,332 @@ export default class CaseNotesController {
+      if (req.query.startDate) queryParams.startDate = req.query.startDate as string
+      if (req.query.endDate) queryParams.endDate = req.query.endDate as string
+      if (req.query.showAll) queryParams.showAll = Boolean(req.query.showAll)
+
+      // Get prisoner data for banner and for use in alerts generation
+      const prisonerData = await this.prisonerSearchService.getPrisonerDetails(clientToken, req.params.prisonerNumber)
+
+      // Set role based permissions
+      const canDeleteSensitiveCaseNotes = userHasRoles([Role.DeleteSensitiveCaseNotes], res.locals.user.userRoles)
+
+      // If user cannot view this prisoner's case notes, redirect to 404 page
+      if (!canViewCaseNotes(res.locals.user, prisonerData)) {
+        return res.render('notFound.njk', {
+          url: req.headers.referer || `/prisoner/${prisonerData.prisonerNumber}`,
+        })
+      }
+
+      let addCaseNoteLinkUrl: string
+      if (canAddCaseNotes(res.locals.user, prisonerData)) {
+        addCaseNoteLinkUrl = `/prisoner/${prisonerData.prisonerNumber}/add-case-note`
+      }
+
+      // Get total count of case notes ignoring filters
+      const caseNotesUsage = await this.prisonApiClientBuilder(clientToken).getCaseNotesUsage(req.params.prisonerNumber)
+      const hasCaseNotes = Array.isArray(caseNotesUsage) && caseNotesUsage.length
+
+      // Get case notes based on given query params
+      const caseNotesPageData = await this.caseNotesService.get(
+        userToken,
+        prisonerData,
+        queryParams,
+        canDeleteSensitiveCaseNotes,
+      )
+
+      const { types, subTypes, typeSubTypeMap } = this.mapCaseNoteTypes(
+        caseNotesPageData.caseNoteTypes,
+        queryParams.type,
+      )
+
+      const showingAll = queryParams.showAll
+
+      // Get staffId to use in conditional logic for amend link
+      const { staffId } = res.locals.user
+
+      // Render page
+      return res.render('pages/caseNotes/caseNotesPage', {
+        pageTitle: 'Case notes',
+        ...mapHeaderData(prisonerData, res.locals.user, 'case-notes'),
+        ...caseNotesPageData,
+        types,
+        subTypes,
+        typeSubTypeMap,
+        showingAll,
+        hasCaseNotes,
+        addCaseNoteLinkUrl,
+        staffId,
+      })
+    }
+  }
+
+  public displayAddCaseNote(): RequestHandler {
+    return async (req: Request, res: Response) => {
+      const { clientToken } = res.locals
+      const userToken = res.locals.user.token
+      const { firstName, lastName, prisonerNumber, prisonId, restrictedPatient } =
+        await this.prisonerSearchService.getPrisonerDetails(clientToken, req.params.prisonerNumber)
+      const prisonerDisplayName = formatName(firstName, undefined, lastName, { style: NameFormatStyle.firstLast })
+
+      // If user cannot view this prisoner's case notes, redirect to 404 page
+      if (!canViewCaseNotes(res.locals.user, { prisonId, restrictedPatient } as Prisoner)) {
```

Comment:
Worrying that this is having to be cast. Perhaps the function arg should receive a partial rather than the full Prisoner if it is not all needed